src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -103,21 +104,28 @@ public ImmutableArray<PackageSource> GetPackageSources()
                 return packageSources;
             }
 
-            try
+            lock (_gate)
             {
-                packageSources = _packageSourceProvider.Value.GetSources(includeUnOfficial: true, includeDisabled: false)
-                    .SelectAsArray(r => new PackageSource(r.Key, r.Value));
+                if (_packageSourcesAsync is null)
+                {
+                    _packageSourcesAsync = ThreadingContext.JoinableTaskFactory.RunAsync(() => GetPackageSourcesImplAsync());
+                }
             }
-            catch (Exception ex) when (ex is InvalidDataException || ex is InvalidOperationException)
+
+            if (allowSwitchToMainThread)
             {
-                // These exceptions can happen when the nuget.config file is broken.
-                packageSources = ImmutableArray<PackageSource>.Empty;
+                using var combinedToken = _tokenSource.Token.CombineWith(cancellationToken);
+                packageSources = await _packageSourcesAsync.JoinAsync(combinedToken.Token).ConfigureAwait(false);
```

Comment:
If we're allowing a switch to the main thread but a workspace change happens because the UI thread isn't blocked, can't this result in the request being cancelled because _tokenSource.Token will be cancelled, even though the caller won't expect it?

Reply:
A agree we shouldn't be cancelling the operation except when `cancellationToken` is canceled. But what do we do for this case?

Reply:
This is now fixed.